File: packages/wasm-solana/js/instructions.ts
/**
 * Instruction decoders for Solana native programs.
 *
 * Provides decoders for:
 * - System Program (transfers, account creation, nonce operations)
 * - Stake Program (staking operations)
 * - ComputeBudget Program (fee and compute limit settings)
 *
 * Note: The underlying WASM decoder supports all instruction types from official
 * Solana crates. TypeScript types are provided for commonly used instructions.
 */

import {
  SystemInstructionDecoder as WasmSystemDecoder,
  StakeInstructionDecoder as WasmStakeDecoder,
  ComputeBudgetInstructionDecoder as WasmComputeBudgetDecoder,
} from "./wasm/wasm_solana.js";

// =============================================================================
// System Instruction Types (commonly used in BitGoJS)
// =============================================================================

export interface SystemCreateAccount {
  type: "CreateAccount";
  lamports: bigint;
  space: bigint;
  owner: string;
}

export interface SystemTransfer {
  type: "Transfer";
  lamports: bigint;
}

export interface SystemAdvanceNonceAccount {
  type: "AdvanceNonceAccount";
}

export interface SystemInitializeNonceAccount {
  type: "InitializeNonceAccount";
  authorized: string;
}

/** Union of commonly used System instruction types */
export type SystemInstruction =
  | SystemCreateAccount
  | SystemTransfer
  | SystemAdvanceNonceAccount
  | SystemInitializeNonceAccount
  | { type: string; [key: string]: unknown }; // Other instruction types

// =============================================================================
// Stake Instruction Types (commonly used in BitGoJS)
// =============================================================================

export interface StakeLockup {
  unixTimestamp: bigint;
  epoch: bigint;
  custodian: string;
}

export interface StakeInitialize {
  type: "Initialize";
  staker: string;
  withdrawer: string;
  lockup: StakeLockup;
}

export interface StakeAuthorize {
  type: "Authorize";
  newAuthority: string;
  stakeAuthorize: "Staker" | "Withdrawer";
}

export interface StakeDelegateStake {
  type: "DelegateStake";
}

export interface StakeSplit {
  type: "Split";
  lamports: bigint;
}

export interface StakeWithdraw {
  type: "Withdraw";
  lamports: bigint;
}

export interface StakeDeactivate {
  type: "Deactivate";
}

export interface StakeMerge {
  type: "Merge";
}

/** Union of commonly used Stake instruction types */
export type StakeInstruction =
  | StakeInitialize
  | StakeAuthorize
  | StakeDelegateStake
  | StakeSplit
  | StakeWithdraw
  | StakeDeactivate
  | StakeMerge
  | { type: string; [key: string]: unknown }; // Other instruction types

// =============================================================================
// ComputeBudget Instruction Types
// =============================================================================

export interface ComputeBudgetSetComputeUnitLimit {
  type: "SetComputeUnitLimit";
  units: number;
}

export interface ComputeBudgetSetComputeUnitPrice {
  type: "SetComputeUnitPrice";
  microLamports: bigint;
}

/** Union of commonly used ComputeBudget instruction types */
export type ComputeBudgetInstruction =
  | ComputeBudgetSetComputeUnitLimit
  | ComputeBudgetSetComputeUnitPrice
  | { type: string; [key: string]: unknown }; // Other instruction types

// =============================================================================
// System Instruction Decoder
// =============================================================================

/** System Program ID */
export const SYSTEM_PROGRAM_ID = "11111111111111111111111111111111";

/**
 * Check if a program ID is the System Program
 * @param programId - The program ID to check (base58 string)
 */
export function isSystemProgram(programId: string): boolean {
  return WasmSystemDecoder.is_system_program(programId);
}

/**
 * Decode a System program instruction from raw bytes.
 * Supports all System instruction types via official Solana crates.
 * @param data - The instruction data bytes
 * @returns The decoded instruction with type discriminant
 * @throws Error if the instruction cannot be decoded
 */
export function decodeSystemInstruction(data: Uint8Array): SystemInstruction {
  return WasmSystemDecoder.decode(data) as SystemInstruction;
}

// =============================================================================
// Stake Instruction Decoder
// =============================================================================

/** Stake Program ID */
export const STAKE_PROGRAM_ID = "Stake11111111111111111111111111111111111111";

/**
 * Check if a program ID is the Stake Program
 * @param programId - The program ID to check (base58 string)
 */
export function isStakeProgram(programId: string): boolean {
  return WasmStakeDecoder.is_stake_program(programId);
}

/**
 * Decode a Stake program instruction from raw bytes.
 * Supports all Stake instruction types via official Solana crates.
 * @param data - The instruction data bytes
 * @returns The decoded instruction with type discriminant
 * @throws Error if the instruction cannot be decoded
 */
export function decodeStakeInstruction(data: Uint8Array): StakeInstruction {
  return WasmStakeDecoder.decode(data) as StakeInstruction;
}

// =============================================================================
// ComputeBudget Instruction Decoder
// =============================================================================

/** ComputeBudget Program ID */
export const COMPUTE_BUDGET_PROGRAM_ID =
  "ComputeBudget111111111111111111111111111111";

/**
 * Check if a program ID is the ComputeBudget Program
 * @param programId - The program ID to check (base58 string)
 */
export function isComputeBudgetProgram(programId: string): boolean {
  return WasmComputeBudgetDecoder.is_compute_budget_program(programId);
}

/**
 * Decode a ComputeBudget program instruction from raw bytes.
 * Supports all ComputeBudget instruction types via official Solana crates.
 * @param data - The instruction data bytes
 * @returns The decoded instruction with type discriminant
 * @throws Error if the instruction cannot be decoded
 */
export function decodeComputeBudgetInstruction(
  data: Uint8Array
): ComputeBudgetInstruction {
  return WasmComputeBudgetDecoder.decode(data) as ComputeBudgetInstruction;
}

File: packages/wasm-solana/src/instructions.rs
//! Instruction decoders using official Solana interface crates.
//!
//! This module wraps official Solana instruction types for WASM compatibility:
//! - `solana-system-interface` for System program
//! - `solana-stake-interface` for Stake program
//! - `solana-compute-budget-interface` for ComputeBudget program

use crate::error::WasmSolanaError;

// Re-export official instruction types
pub use solana_compute_budget_interface::ComputeBudgetInstruction;
pub use solana_stake_interface::instruction::StakeInstruction;
pub use solana_system_interface::instruction::SystemInstruction;

/// Program IDs as base58 strings
pub const SYSTEM_PROGRAM_ID: &str = "11111111111111111111111111111111";
pub const STAKE_PROGRAM_ID: &str = "Stake11111111111111111111111111111111111111";
pub const COMPUTE_BUDGET_PROGRAM_ID: &str = "ComputeBudget111111111111111111111111111111";

/// Decode a System program instruction from raw bytes.
pub fn decode_system_instruction(data: &[u8]) -> Result<SystemInstruction, WasmSolanaError> {
    bincode::deserialize(data)
        .map_err(|e| WasmSolanaError::new(&format!("Failed to decode System instruction: {}", e)))
}

/// Decode a Stake program instruction from raw bytes.
pub fn decode_stake_instruction(data: &[u8]) -> Result<StakeInstruction, WasmSolanaError> {
    bincode::deserialize(data)
        .map_err(|e| WasmSolanaError::new(&format!("Failed to decode Stake instruction: {}", e)))
}

/// Decode a ComputeBudget program instruction from raw bytes.
pub fn decode_compute_budget_instruction(
    data: &[u8],
) -> Result<ComputeBudgetInstruction, WasmSolanaError> {
    use borsh::BorshDeserialize;
    ComputeBudgetInstruction::try_from_slice(data)
        .map_err(|e| WasmSolanaError::new(&format!("Failed to decode ComputeBudget instruction: {}", e)))
}

/// Check if a program ID is the System program.
pub fn is_system_program(program_id: &str) -> bool {
    program_id == SYSTEM_PROGRAM_ID
}

/// Check if a program ID is the Stake program.
pub fn is_stake_program(program_id: &str) -> bool {
    program_id == STAKE_PROGRAM_ID
}

/// Check if a program ID is the ComputeBudget program.
pub fn is_compute_budget_program(program_id: &str) -> bool {
    program_id == COMPUTE_BUDGET_PROGRAM_ID
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_decode_system_transfer() {
        // Transfer 100000 lamports (discriminator 2 + u64 lamports)
        let data = [
            2, 0, 0, 0, // discriminator = 2 (Transfer)
            160, 134, 1, 0, 0, 0, 0, 0, // lamports = 100000
        ];

        let instr = decode_system_instruction(&data).unwrap();
        match instr {
            SystemInstruction::Transfer { lamports } => {
                assert_eq!(lamports, 100000);
            }
            _ => panic!("Expected Transfer instruction"),
        }
    }

    #[test]
    fn test_decode_system_advance_nonce() {
        let data = [4, 0, 0, 0]; // discriminator = 4 (AdvanceNonceAccount)
        let instr = decode_system_instruction(&data).unwrap();
        assert!(matches!(instr, SystemInstruction::AdvanceNonceAccount));
    }

    #[test]
    fn test_decode_stake_delegate() {
        let data = [2, 0, 0, 0]; // discriminator = 2 (DelegateStake)
        let instr = decode_stake_instruction(&data).unwrap();
        assert!(matches!(instr, StakeInstruction::DelegateStake));
    }

    #[test]
    fn test_decode_stake_deactivate() {
        let data = [5, 0, 0, 0]; // discriminator = 5 (Deactivate)
        let instr = decode_stake_instruction(&data).unwrap();
        assert!(matches!(instr, StakeInstruction::Deactivate));
    }

    #[test]
    fn test_decode_compute_budget_set_limit() {
        let data = [
            2, // discriminator = 2 (SetComputeUnitLimit)
            64, 66, 15, 0, // units = 1000000
        ];
        let instr = decode_compute_budget_instruction(&data).unwrap();
        assert!(matches!(
            instr,
            ComputeBudgetInstruction::SetComputeUnitLimit(1000000)
        ));
    }

    #[test]
    fn test_decode_compute_budget_set_price() {
        let data = [
            3, // discriminator = 3 (SetComputeUnitPrice)
            232, 3, 0, 0, 0, 0, 0, 0, // micro_lamports = 1000
        ];
        let instr = decode_compute_budget_instruction(&data).unwrap();
        assert!(matches!(
            instr,
            ComputeBudgetInstruction::SetComputeUnitPrice(1000)
        ));
    }

    #[test]
    fn test_program_id_checks() {
        assert!(is_system_program(SYSTEM_PROGRAM_ID));
        assert!(!is_system_program(STAKE_PROGRAM_ID));
        assert!(is_stake_program(STAKE_PROGRAM_ID));
        assert!(is_compute_budget_program(COMPUTE_BUDGET_PROGRAM_ID));
    }
}
